pkg/certgenerator/v1beta1/generator.go

```diff
@@ -0,0 +1,264 @@
+		return fmt.Errorf("%w: %v", errServiceNotFound, err)
+	}
+
+	certExist, err := c.isCertExist(ctx)
+	if err != nil {
+		return fmt.Errorf("%w: %v", errCertCheckFail, err)
+	}
+	if !certExist {
+		c.fullServiceDomain = strings.Join([]string{c.serviceName, c.namespace, "svc"}, ".")
+
+		if err = c.createCert(); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertFail, err)
+		}
+		if err = c.updateCertSecret(ctx); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertSecretFail, err)
+		}
+	}
+	if err = c.injectCert(ctx); err != nil {
+		return fmt.Errorf("%w: %v", errInjectCertError, err)
+	}
+	return nil
+}
+
+// isCertExist checks if a secret embedded certs already exists.
+// For example, it will return true if the katib-controller is created with enabled leader-election
+// since another controller pod will create the secret.
+func (c *CertGenerator) isCertExist(ctx context.Context) (bool, error) {
+	secret := &corev1.Secret{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.secretName, Namespace: c.namespace}, secret); err != nil {
+		return false, err
+	}
+	key := secret.Data[serverKeyName]
+	cert := secret.Data[serverCertName]
+	if len(key) != 0 && len(cert) != 0 {
+		c.certs = &certificates{
+			keyPem:  key,
+			certPem: cert,
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// createCert creates the self-signed certificate and private key.
+func (c *CertGenerator) createCert() error {
+	now := time.Now()
+	template := &x509.Certificate{
+		SerialNumber: big.NewInt(0),
+		Subject: pkix.Name{
+			CommonName: c.fullServiceDomain,
+		},
+		DNSNames: []string{
+			c.fullServiceDomain,
+		},
+		NotBefore:   now,
+		NotAfter:    now.Add(24 * time.Hour * 365 * 10),
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+	}
+
+	klog.Info("Generating self-signed public certificate and private key.")
+	rawKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+
+	der, err := x509.CreateCertificate(rand.Reader, template, template, rawKey.Public(), rawKey)
+	if err != nil {
+		return err
+	}
+	if c.certs, err = encode(rawKey, der); err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateCertSecret updates Secret embedded tls.key and tls.crt.
+func (c *CertGenerator) updateCertSecret(ctx context.Context) error {
+	secret := &corev1.Secret{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.secretName, Namespace: c.namespace}, secret); err != nil {
+		return err
+	}
+	newSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			UID:        secret.UID,
+			Name:       secret.Name,
+			Namespace:  secret.Namespace,
+			Generation: secret.Generation,
+		},
+		TypeMeta: secret.TypeMeta,
+	}
+	if len(newSecret.APIVersion) == 0 {
+		newSecret.APIVersion = corev1.SchemeGroupVersion.String()
+	}
+	if len(newSecret.Kind) == 0 {
+		newSecret.Kind = "Secret"
+	}
```

Comment:
In the SSA, the `newSecret` is a PATCH. So we shouldn't copy all fields to the PATCH to reduce calculating cost.

Reply:
In particular, we should carefully handle `.metadata. managedFields` and `.metadata.resourceVersion`.

Reply:
@tenzen-y Won't Name and Namespace be sufficient for SSA ?
[Looking to the docs](https://kubernetes.io/docs/reference/using-api/server-side-apply/#transferring-ownership), we need only name and namespace for the patch.

Reply:
@andreyvelich I think we need UID and Generation. Because `1. if others accidentally update UID` and `2. if others update Secret and Generation is updated`,  we can produce a conflict and then force override the fields by passing UID and Generation.

Reply:
@tenzen-y What do you mean by others ? Other controller pods ?

Reply:
I meant others are `other controller` or users (human).
However, more investigation showed me that `generation` doesn't affect the Secret.
@andreyvelich So, I think we need the below field for the patch. Any thoughts?

- ValidatingWebhookConfiguraion and MutatingWebhookConfiguration: Name, Namespace, UID, and Generation.
- Secret: Name, Namespace, and UID

Reply:
@tenzen-y Makes sense, so I believe we can also r[emove the APIVersion and Kind check](https://github.com/kubeflow/katib/blob/7ef511fbdc8d761dc4a7adad424c274b193af90f/pkg/certgenerator/v1beta1/generator.go#L246-L251), right ?
It is not possible to check those for existing Kubernetes objects.

Reply:
@andreyvelich It makes sense. I will remove the cheks.